src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -22,121 +22,194 @@
     private static final double RANGE_MERGING_PREFERENCE = 1.5;
 
     private String mbeanName;
-    private boolean registered = false;
-
-    private volatile HashMap<InetAddressAndPort, Double> scores = new HashMap<>();
-    private final ConcurrentHashMap<InetAddressAndPort, ExponentiallyDecayingReservoir> samples = new ConcurrentHashMap<>();
+    private boolean mbeanRegistered = false;
 
     public final IEndpointSnitch subsnitch;
 
-    private volatile ScheduledFuture<?> updateSchedular;
-    private volatile ScheduledFuture<?> resetSchedular;
+    private volatile ScheduledFuture<?> updateScheduler;
+    private volatile ScheduledFuture<?> latencyProbeScheduler;
 
-    private final Runnable update;
-    private final Runnable reset;
+    private final Runnable update = this::updateScores;
+    private final Runnable latencyProbe = this::maybeSendLatencyProbe;
 
     public DynamicEndpointSnitch(IEndpointSnitch snitch)
     {
         this(snitch, null);
     }
 
-    public DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
+    protected DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
     {
         mbeanName = "org.apache.cassandra.db:type=DynamicEndpointSnitch";
         if (instance != null)
             mbeanName += ",instance=" + instance;
         subsnitch = snitch;
-        update = new Runnable()
-        {
-            public void run()
-            {
-                updateScores();
-            }
-        };
-        reset = new Runnable()
-        {
-            public void run()
-            {
-                // we do this so that a host considered bad has a chance to recover, otherwise would we never try
-                // to read from it, which would cause its score to never change
-                reset();
-            }
-        };
 
         if (DatabaseDescriptor.isDaemonInitialized())
         {
-            updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
-            resetSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(reset, dynamicResetInterval, dynamicResetInterval, TimeUnit.MILLISECONDS);
+            applyConfigChanges();
             registerMBean();
         }
+        nextProbeGenerationTime = System.nanoTime();
+    }
+
+    /**
+     * Records a latency. This MUST be cheap as it is called in the fast path
+     */
+    abstract public void receiveTiming(InetAddressAndPort address, long latency, LatencyMeasurementType measurementType);
+
+    /**
+     * Called periodically from {@link DynamicEndpointSnitch#updateScores()}. This method returns a new score map
+     * that will be assigned to the {@link DynamicEndpointSnitch#scores} map.
+     * This is generally more expensive than {@link DynamicEndpointSnitch#receiveTiming}
+     */
+    abstract public Map<InetAddressAndPort, Double> calculateScores();
```

Comment:
I was thinking that it would be pretty different in future implementations, but I can just make it protected and then subclasses can redefine it anyways :-)

Reply:
So I think future implementations just override the superclass method if they want?

Reply:
Additionally, the might not end up being the super class for future snitch implementations since they can implement IEndpointSnitch and replace whatever they want. All we have to worry about today is solving for EMA and Histogram I think.